src/Cryptol/Eval/SBV.hs

```diff
@@ -0,0 +1,767 @@
+
+  wordSignedLessThan _ x y = pure $! svLessThan sx sy
+    where sx = svSign x
+          sy = svSign y
+
+  joinWord _ x y = pure $! svJoin x y
+
+  splitWord _ _leftW rightW w = pure
+    ( svExtract (intSizeOf w - 1) (fromInteger rightW) w
+    , svExtract (fromInteger rightW - 1) 0 w
+    )
+
+  extractWord _ len start w =
+    pure $! svExtract (fromInteger start + fromInteger len - 1) (fromInteger start) w
+
+  wordAnd _ a b = pure $! svAnd a b
+  wordOr  _ a b = pure $! svOr a b
+  wordXor _ a b = pure $! svXOr a b
+  wordComplement _ a = pure $! svNot a
+
+  wordPlus  _ a b = pure $! svPlus a b
+  wordMinus _ a b = pure $! svMinus a b
+  wordMult  _ a b = pure $! svTimes a b
+  wordNegate _ a  = pure $! svUNeg a
+
+  wordDiv sym a b =
+    do let z = literalSWord (intSizeOf b) 0
+       assertSideCondition sym (svNot (svEqual b z)) DivideByZero
+       pure $! svQuot a b
+
+  wordMod sym a b =
+    do let z = literalSWord (intSizeOf b) 0
+       assertSideCondition sym (svNot (svEqual b z)) DivideByZero
+       pure $! svRem a b
+
+  wordSignedDiv sym a b =
+    do let z = literalSWord (intSizeOf b) 0
+       assertSideCondition sym (svNot (svEqual b z)) DivideByZero
+       pure $! signedQuot a b
+
+  wordSignedMod sym a b =
+    do let z = literalSWord (intSizeOf b) 0
+       assertSideCondition sym (svNot (svEqual b z)) DivideByZero
+       pure $! signedRem a b
+
+  wordExp _ a b = sExp a b
+  wordLg2 _ a = sLg2 a
+
+  wordToInt _ x = pure $! svToInteger x
+  wordFromInt _ w i = pure $! svFromInteger w i
+
+  intEq _ a b = pure $! svEqual a b
+  intLessThan _ a b = pure $! svLessThan a b
+  intGreaterThan _ a b = pure $! svGreaterThan a b
+
+  intPlus  _ a b = pure $! svPlus a b
+  intMinus _ a b = pure $! svMinus a b
+  intMult  _ a b = pure $! svTimes a b
+  intNegate _ a  = pure $! SBV.svUNeg a
+  intLg2 _ a = svLg2 a
+
+  intDiv sym a b =
+    do let z = svInteger KUnbounded 0
+       assertSideCondition sym (svNot (svEqual b z)) DivideByZero
+       pure $! svQuot a b  -- TODO! Fix this: see issue #662
+  intMod sym a b =
+    do let z = svInteger KUnbounded 0
+       assertSideCondition sym (svNot (svEqual b z)) DivideByZero
+       pure $! svRem a b   -- TODO! Fix this: see issue #662
+
+  intExp sym a b
+    | Just e <- svAsInteger b =
+       do unless (0 <= e) (raiseError sym NegativeExponent)
+          pure $! SBV.svExp a b
+    | otherwise =
+       raiseError sym (UnsupportedSymbolicOp "integer exponentation")
+
+  -- NB, we don't do reduction here
+  intToZn _ _m a = pure a
+
+  znToInt _ 0 a = pure a
+  znToInt _ m a =
+    do let m' = svInteger KUnbounded m
+       pure $! svRem a m'
+
+  znEq _ 0 a b = pure $! svEqual a b
+  znEq _ m a b = svDivisible m (SBV.svMinus a b)
+
+  znPlus  _ m a b = sModAdd m a b
+  znMinus _ m a b = sModSub m a b
+  znMult  _ m a b = sModMult m a b
+  znNegate _ m a  = sModNegate m a
+
+-- TODO: implement this properly in SBV using "bv2int"
```

Comment:
Is this an easy thing to do? Or something we should have an issue for?

Reply:
I remember chatting with Levent about this. Some solvers support `bv2int` and/or `int2bv` functions (I don't think they're standardized in SMT-Lib) but apparently they're often implemented incompletely: Basically, you might get some rewrite rules applied in the external solver and then afterward they're treated as uninterpreted functions. So using `bv2int`/`int2bv` can lead to bogus counterexamples. Levent preferred to avoid using them.

Reply:
We should just remove the comment. It turns out there actually was an sbv ticket for this: LeventErkok/sbv#233 (closed after deciding not to implement it)

Reply:
I'll remove the comments